### PR TITLE
IBX-4001: Made `ezimagefile` entries unique based on `fieldId` and `path`

### PR DESCRIPTION
**Pull request by barw4**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-4001](https://issues.ibexa.co/browse/IBX-4001)
| **Type**                                   | bug
| **Target Ibexa version** | `v3.3`
| **BC breaks**                          | no

Following PR prevents from flooding an `ezimagefile` table with duplicated records.

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/engineering-team`).

### CODE REVIEW

File: eZ/Publish/Core/FieldType/Image/ImageStorage.php [hunk 1]
```diff
@@ -128,9 +132,6 @@ public function storeFieldData(VersionInfo $versionInfo, Field $field, array $co
             $field->value->externalData = null;
         }
 
-        $this->gateway->storeImageReference($field->value->data['uri'], $field->id);
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Image/ImageStorage.php:131]: Hmm, it's ok to add extra if, but why did you move this to if-else block above?

Comment (barw4) [reply eZ/Publish/Core/FieldType/Image/ImageStorage.php:131]: If `externalData` is empty I think we don't need to store anything in an `ezimagefile` table

Comment (alongosz) [reply eZ/Publish/Core/FieldType/Image/ImageStorage.php:131]: > If `externalData` is empty I think we don't need to store anything in an `ezimagefile` table

Hmm... but here you operate on `value->data` not `value->externalData`. Does `externalData` imply `data` to be empty?

**Review by alongosz — COMMENTED** [review]
Replied via https://github.com/ezsystems/ezplatform-kernel/pull/343#discussion_r996943915

File: eZ/Publish/Core/FieldType/Image/ImageStorage.php (hunk 1 above)
Comment (barw4) [reply eZ/Publish/Core/FieldType/Image/ImageStorage.php:131]: https://github.com/ezsystems/ezplatform-kernel/blob/86a6a8e7119488c63ccfbde280483a7c03c8aed4/eZ/Publish/Core/FieldType/Image/ImageStorage.php#L108-L111
Yes, but `value->data`, that is built from `externalData`, is taken to this `if` directly from the lines above. Besides, we also have a comment (https://github.com/ezsystems/ezplatform-kernel/blob/86a6a8e7119488c63ccfbde280483a7c03c8aed4/eZ/Publish/Core/FieldType/Image/ImageStorage.php#L118) saying that this is not a new image.

Comment (barw4) [reply eZ/Publish/Core/FieldType/Image/ImageStorage.php:131]: I mean, `externalData` doesn't imply `data` to be empty but it implies we want to store something in the mentioned table.

File: eZ/Publish/Core/FieldType/Image/ImageStorage/Gateway/DoctrineStorage.php [hunk 2]
```diff
@@ -456,4 +456,38 @@ public function updateImagePath(int $fieldId, string $oldPath, string $newPath):
             ->execute()
         ;
     }
+
+    /**
+     * @throws \eZ\Publish\Core\IO\Exception\InvalidBinaryFileIdException
+     * @throws \Doctrine\DBAL\Driver\Exception
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function getImageReference(string $uri, int $fieldId): bool
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Image/ImageStorage/Gateway/DoctrineStorage.php]: naming: you're saying `get` image reference, yet your returning bool. It seems you mean `hasImageReference`, right?

File: eZ/Publish/Core/FieldType/Image/ImageStorage/Gateway/DoctrineStorage.php [hunk 3]
```diff
@@ -456,4 +456,38 @@ public function updateImagePath(int $fieldId, string $oldPath, string $newPath):
...
+            ->where(
+                $selectQuery->expr()->eq(
+                    $this->connection->quoteIdentifier('filepath'),
+                    ':path'
+                )
+            )
+            ->andWhere(
+                $selectQuery->expr()->eq(
+                    $this->connection->quoteIdentifier('contentobject_attribute_id'),
+                    ':fieldId'
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Image/ImageStorage/Gateway/DoctrineStorage.php]: let's use here Doctrine naming convention. Whenever we use camel case for QB parameter names, it's technical debt.
```suggestion
                    ':field_id'
```

File: eZ/Publish/Core/FieldType/Image/ImageStorage/Gateway/DoctrineStorage.php [hunk 4]
```diff
@@ -456,4 +456,38 @@ public function updateImagePath(int $fieldId, string $oldPath, string $newPath):
...
+                )
+            )
+            ->setParameter(':path', $path)
+            ->setParameter(':fieldId', $fieldId)
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Image/ImageStorage/Gateway/DoctrineStorage.php]:
```suggestion
            ->setParameter(':field_id', $fieldId)
```

File: tests/integration/Core/Image/ImageStorage/ImageStorageTest.php [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+                    $this->gateway,
+                    $this->ioService,
+                    $this->pathGenerator,
+                    $this->imageSizeMetadataHandler,
+                    $this->deprecationWarner,
+                    $this->aliasCleaner,
+                    $this->filePathNormalizer,
+                ]
+            )
+            ->getMock();
```
Comment (alongosz) [inline tests/integration/Core/Image/ImageStorage/ImageStorageTest.php]: Rule `#1` of writing unit tests: do not mock a class under test :-)

This should work fine:
```php
new ImageStorage(
    $this->gateway,
    $this->ioService,
    $this->pathGenerator,
    $this->imageSizeMetadataHandler,
    $this->deprecationWarner,
    $this->aliasCleaner,
    $this->filePathNormalizer,
);
```

File: tests/integration/Core/Image/ImageStorage/ImageStorageTest.php [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        $this->ioService
+            ->expects(self::once())
+            ->method('getMimeType')
+            ->with($binaryFile->id)
+            ->willReturn('image/jpeg');
+
+        return $binaryFile;
+    }
+
+    public function providerOfFieldData(): array
```
Comment (alongosz) [inline tests/integration/Core/Image/ImageStorage/ImageStorageTest.php]: new[1] fashion:
```suggestion
    public function providerOfFieldData(): iterable
```

[1] actually a few years old

File: tests/integration/Core/Image/ImageStorage/ImageStorageTest.php [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            'contentInfo' => new ContentInfo([
+                'id' => 236,
+                'contentTypeId' => 25,
+            ]),
+            'versionNo' => 1,
+        ]);
+
+        return [
+            [$versionInfo, $field],
+        ];
```
Comment (alongosz) [inline tests/integration/Core/Image/ImageStorage/ImageStorageTest.php]: new fashion continuation:
```suggestion
        yield [$versionInfo, $field];
```
;-)

**Review by alongosz — CHANGES_REQUESTED** [review]

**Review by alongosz — APPROVED** [review]

File: tests/integration/Core/Image/ImageStorage/ImageStorageTest.php [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+use eZ\Publish\Core\IO\UrlRedecoratorInterface;
+use eZ\Publish\Core\IO\Values\BinaryFile;
+use eZ\Publish\Core\IO\Values\BinaryFileCreateStruct;
+use eZ\Publish\SPI\Persistence\Content\Field;
+use eZ\Publish\SPI\Persistence\Content\FieldValue;
+use eZ\Publish\SPI\Persistence\Content\VersionInfo;
+
+final class ImageStorageTest extends BaseCoreFieldTypeIntegrationTest
+{
+    /** @var \eZ\Publish\Core\FieldType\Image\ImageStorage\Gateway|\PHPUnit\Framework\MockObject\MockObject */
```
Comment (konradoboza) [inline tests/integration/Core/Image/ImageStorage/ImageStorageTest.php]:
```suggestion
    /** @var \eZ\Publish\Core\FieldType\Image\ImageStorage\Gateway */
```

File: tests/integration/Core/Image/ImageStorage/ImageStorageTest.php [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        $this->ioService
+            ->expects(self::once())
+            ->method('getMimeType')
+            ->with($binaryFile->id)
+            ->willReturn('image/jpeg');
+
+        return $binaryFile;
+    }
+
+    public function providerOfFieldData(): iterable
```
Comment (konradoboza) [inline tests/integration/Core/Image/ImageStorage/ImageStorageTest.php:215]:
```suggestion
    /**
     * @return iterable<array{
     *     eZ\Publish\SPI\Persistence\Content\VersionInfo,
     *     eZ\Publish\SPI\Persistence\Content\Field
     * }>
     */
    public function providerOfFieldData(): iterable
```

File: tests/integration/Core/Image/ImageStorage/ImageStorageTest.php [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+            $this->filePathNormalizer,
+        );
+    }
+
+    public function testHasFieldData(): void
+    {
+        self::assertTrue($this->storage->hasFieldData());
+    }
+
+    private function getContext(): array
```
Comment (konradoboza) [inline tests/integration/Core/Image/ImageStorage/ImageStorageTest.php:88]:
```suggestion
    /**
     * @return array<string, string>
     */
    private function getContext(): array
```

File: eZ/Publish/Core/FieldType/Image/ImageStorage/Gateway/DoctrineStorage.php [hunk 11]
```diff
@@ -456,4 +456,38 @@ public function updateImagePath(int $fieldId, string $oldPath, string $newPath):
...
+     */
+    public function hasImageReference(string $uri, int $fieldId): bool
+    {
+        $path = $this->redecorator->redecorateFromSource($uri);
+
+        $selectQuery = $this->connection->createQueryBuilder();
+        $selectQuery
+            ->select(1)
+            ->from($this->connection->quoteIdentifier(self::IMAGE_FILE_TABLE))
+            ->where(
```
Comment (Steveb-p) [inline eZ/Publish/Core/FieldType/Image/ImageStorage/Gateway/DoctrineStorage.php]: `where` `QueryBuilder`'s method should be avoided, because it has a side-effect of discarding any previous where statements that might have been added. I know that is not the case here (there are none), but in general Doctrine recommends using only `andWhere`.
```suggestion
            ->andWhere(
```

File: tests/integration/Core/Image/ImageStorage/ImageStorageTest.php [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+        $this->redecorator
+            ->expects(self::exactly(3))
+            ->method('redecorateFromSource')
+            ->with($binaryFile->uri)
+            ->willReturn($binaryFile->uri);
+
+        $this->storage->storeFieldData($versionInfo, $field, $this->getContext());
+
+        self::assertEquals(1, $this->gateway->countImageReferences($binaryFile->uri));
```
Comment (Steveb-p) [inline tests/integration/Core/Image/ImageStorage/ImageStorageTest.php]: Prefer strict equality checks.
```suggestion
        self::assertSame(1, $this->gateway->countImageReferences($binaryFile->uri));
```

File: tests/integration/Core/Image/ImageStorage/ImageStorageTest.php [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+        $this->redecorator
+            ->expects(self::exactly(2))
+            ->method('redecorateFromSource')
+            ->with($binaryFile->uri)
+            ->willReturn($binaryFile->uri);
+
+        $this->storage->storeFieldData($versionInfo, $field, $this->getContext());
+
+        self::assertEquals(1, $this->gateway->countImageReferences($binaryFile->uri));
```
Comment (Steveb-p) [inline tests/integration/Core/Image/ImageStorage/ImageStorageTest.php]:
```suggestion
        self::assertSame(1, $this->gateway->countImageReferences($binaryFile->uri));
```

File: tests/integration/Core/Image/ImageStorage/ImageStorageTest.php [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    /**
+     * @dataProvider providerOfFieldData
+     *
+     * @depends      testStoreFieldDataDuringCreate
+     */
+    public function testStoreFieldDataDuringUpdate(VersionInfo $versionInfo, Field $field): void
```
Comment (Steveb-p) [inline tests/integration/Core/Image/ImageStorage/ImageStorageTest.php:119]: A test that shows that a reference is added when an image is changed should also be part of the test.

(Currently you're testing that a reference will not be added if image is not changed).

**Review by konradoboza — APPROVED** [review]

**Review by Steveb-p — APPROVED** [review]

**Review by tomaszszopinski — APPROVED** [review]
QA approved on IbexaDXP 3.3 commerce.
